src/target/riscv/riscv.c

```diff
@@ -6031,45 +5798,44 @@ int riscv_init_registers(struct target *target)
 					}
 			}
 
-		} else if (number == GDB_REGNO_PRIV) {
-			sprintf(reg_name, "priv");
+		} else if (reg_num == GDB_REGNO_PRIV) {
 			r->group = "general";
 			r->feature = &feature_virtual;
 			r->size = 8;
 
-		} else if (number >= GDB_REGNO_V0 && number <= GDB_REGNO_V31) {
+		} else if (reg_num >= GDB_REGNO_V0 && reg_num <= GDB_REGNO_V31) {
 			r->caller_save = false;
 			r->exist = (info->vlenb > 0);
 			r->size = info->vlenb * 8;
-			sprintf(reg_name, "v%d", number - GDB_REGNO_V0);
 			r->group = "vector";
 			r->feature = &feature_vector;
 			r->reg_data_type = &info->type_vector;
 
-		} else if (number >= GDB_REGNO_COUNT) {
+		} else if (reg_num >= GDB_REGNO_COUNT) {
 			/* Custom registers. */
+			struct reg_name *additional_regs = info->custom_register_names.reg_names;
+			const unsigned int custom_register_index = reg_num - GDB_REGNO_COUNT;
+
 			assert(!list_empty(&info->expose_custom));
+			assert(custom_register_index < info->custom_register_names.names_num);
 
 			range_list_t *range = list_first_entry(&info->expose_custom, range_list_t, list);
 
-			unsigned custom_number = range->low + custom_within_range;
+			unsigned int custom_number = range->low + custom_within_range;
```

Comment:
This is related to the original code (not to the changes in this merge request):

@timsifive Can you please help me check the computation of `custom_number`. This line does not look right. The use of always the first range (list_first_entry) seems incorrect.

Reply:
I agree this looks wrong. The code should walk the linked list to find the range that actually includes reg_num, and use that base.

But this is a pre-existing bug, so doesn't have to be fixed in this PR (mostly to keep the scope limited). I filed #914 to track it.

Reply:
Thank you for looking at it and recording the issue.

Marking this thread as resolved.

Reply:
This is not, a bug. Please, take a look at [the call  to `list_rotate_left()` on line 5843](
https://github.com/riscv/riscv-openocd/pull/896/files/85cde238b39f38707f08e9ebd4077faf709ecad2#diff-b4aa16f9e42cb8f0934baa7c8e0ec9c70a369bef98b99b26ae2e896c8aa95d6aR5843).